File: DataPrep.R
#'---
#' title: "TSCI 5230: Introduction to Data Science"
#' author: 'Jeff Wang'
#' abstract: |
#'  | Provide a summary of objectives, study design, setting, participants,
#'  | sample size, predictors, outcome, statistical analysis, results,
#'  | and conclusions.
#' documentclass: article
#' description: 'Manuscript'
#' clean: false
#' self_contained: true
#' number_sections: false
#' keep_md: true
#' fig_caption: true
#' output:
#'  html_document:
#'    toc: true
#'    toc_float: true
#'    code_folding: show
#' ---
#'
#+ init, echo=FALSE, message=FALSE, warning=FALSE
# init ----
# This part does not show up in your rendered report, only in the script,
# because we are using regular comments instead of #' comments
debug <- 0;
knitr::opts_chunk$set(echo=debug>-1, warning=debug>0, message=debug>0);

library(ggplot2); # visualisation
library(GGally);
#library(rio);# simple command for importing and exporting
library(pander); # format tables
library(printr); # set limit on number of lines printed
library(broom); # allows to give clean dataset
library(dplyr); #add dplyr library
library(tidyr);
library(purrr);
library(table1);

options(max.print=42);
panderOptions('table.split.table',Inf); panderOptions('table.split.cells',Inf);

# load data ----

if(!file.exists("working_script.rdata")){
  system("R -f data.R")
}

load("working_script.rdata")

View(patients)

##plotting using ggplot
ggplot(data = patients, aes(x = anchor_age, fill = gender))+
  geom_histogram() +
  geom_vline(xintercept = 65)

table(patients$gender)
#check for duplicates in the subject_id

#aggregating data: data structures

Demographics <- group_by(admissions, subject_id)%>%
  mutate(los = difftime(dischtime, admittime))%>%
  summarise(admits = n(),
            ethnic = length(unique(ethnicity)),
            ethnicity_combo = paste(sort(unique(ethnicity)), collapse = ":"),
            # language0 <- length(unique(language)),
            # language_combo <- paste(sort(unique(language)), collapse = ":")
            language = tail(language, 1),
            dod = max(deathtime, na.rm = TRUE),
            los = median(los),
            numED = length(na.omit(edregtime)))


 #subset(ethnic > 1)

table(admissions$ethnicity)



ggplot(data = Demographics, aes(x = admits)) +
  geom_histogram()

intersect(names(Demographics), names(patients))

Demographics$subject_id

intersect(Demographics$subject_id, patients$subject_id)

setdiff(Demographics$subject_id, patients$subject_id)

setdiff(Demographics$dod, patients$dod)

Demographics1 <- left_join(Demographics, select(patients, -dod), by = c("subject_id"))


#patient = subset(patients, select = -c(dod))
#paste0(letters,LETTERS, collapse = '---')

############################Vanc/Zosyn study data
# build list of keywords
kw_abx <- c("vanco", "zosyn", "piperacillin", "tazobactam", "cefepime", "meropenam", "ertapenem", "carbapenem", "levofloxacin")
kw_lab <- c("creatinine")
kw_aki <- c("acute renal failure", "acute kidney injury", "acute kidney failure", "acute kidney", "acute renal insufficiency")
kw_aki_pp <- c("postpartum", "labor and delivery")


# search for those keywords in the tables to find the full label names
# remove post partum from aki in last line here
# may need to remove some of the lab labels as well (pending)
# vertical pipe takes the arguement and uses it as an "or" command for the the list of keywords that is being searched
label_abx <- grep(paste0(kw_abx, collapse = '|'), d_items$label, ignore.case = T, value = T, invert = F)
label_lab <- grep(paste0(kw_lab, collapse = '|'), d_labitems$label, ignore.case = T, value = T, invert = F)
label_aki <- grep(paste0(kw_aki, collapse = '|'), d_icd_diagnoses$long_title, ignore.case = T, value = T, invert = F)
label_aki <- grep(paste0(kw_aki_pp, collapse = '|'), label_aki, ignore.case = T, value = T, invert = T)


# use dplyr filter to make tables with the item_id for the keywords above
item_ids_abx <- d_items %>% filter(label %in% label_abx)
item_ids_lab <- d_labitems %>% filter(label %in% label_lab)
item_ids_aki <- d_icd_diagnoses %>% filter(long_title %in% label_aki)

subset(item_ids_abx, category == 'Antibiotics') #Only selects rows with category of Antibiotics
subset(item_ids_abx, category == 'Antibiotics') %>%
  left_join(inputevents, by = 'itemid') #By using subset first in left_join, starting off
#by only selecting rows with antibiotics, and then pulling inputevents data for those
#patients that received the antibiotics with our specified IDs

Antibiotics <- subset(item_ids_abx, category == 'Antibiotics') %>%
  left_join(inputevents, by = 'itemid')

grep('N17', diagnoses_icd$icd_code, value = T) #ICD codes found within the dataset
grep('^548|^N17', diagnoses_icd$icd_code, value=T) #Either 548... or N17... values
#within the diagnosis_icd$icd_code data set
grepl('^548|^N17', diagnoses_icd$icd_code) #True/False for each row whether it contains value
subset(diagnoses_icd,grepl('^548|^N17',icd_code)) #Pulls only the rows that have ICD code of interest
Akidiagnoses_icd <- subset(diagnoses_icd,grepl('^548|^N17',icd_code))

Cr_labevents <- subset(item_ids_lab, fluid == "Blood") %>%
  left_join(labevents, by = 'itemid') #Filter only blood Cr and match to lab events

grepl(paste(kw_abx, collapse='|'),emar$medication)
subset(emar,grepl(paste(kw_abx, collapse='|'),medication,ignore.case = T))$event_txt%>%
  table()%>%sort() #Filter emar by antibiotic administration with individual event txt

#merge lab events and antibiotic administration

#comparing Vancomycin to Zosyn


Antibiotic_Groupings <- group_by(Antibiotics, hadm_id)%>%
  summarise(Vanc = 'Vancomycin' %in% label, Zosyn = any(grepl('Piperacillin', label)),
            N = n(),
            Other = length(grep('Piperacillin|Vancomycin',
                                label, val = TRUE, invert = TRUE))>0,
            Exposure1 = case_when(!Vanc ~ 'Other',
                                  Vanc&Zosyn ~ 'Vanc & Zosyn',
                                  Other ~ 'Vanc & Other',
                                  !Other ~ 'Vanc',
                                  TRUE ~ 'UNDEFINED'),
            Exposure2 = case_when(Vanc & !Zosyn & !Other ~ 'Vanc',
                                  Vanc & Zosyn & !Other ~ 'Vanc & Zosyn',
                                  Vanc & Zosyn & Other ~ 'Vanc & Zosyn',
                                  !Vanc & !Zosyn ~ 'Other'))
#Vanc & !Zosyn & !Other ~ 'Vanc' example: make another column for exposure 2


Admission_Scaffold <- admissions %>% select(hadm_id, admittime, dischtime) %>%
  transmute(hadm_id = hadm_id,
            ip_date = map2(as.Date(admittime), as.Date(dischtime), seq, by = "1 day"))%>%
  unnest(ip_date)



Antibiotics_dates <- Antibiotics %>%
  transmute(hadm_id = hadm_id,
            group = case_when('Vancomycin' == label ~ "Vanc",
                              grepl('Piperacillin', label) ~ "Zosyn",
                              TRUE ~ "Other"),
            starttime = starttime,
            endtime = endtime)%>%
  unique()%>%
  subset(!is.na(starttime) & !is.na(endtime))%>%
  transmute(hadm_id = hadm_id,
            ip_date = {oo <- try(map2(as.Date(starttime), as.Date(endtime), seq, by = "1 day"));
            if(is(oo, 'try-error')){browser()}
            oo},
            group = group)%>%
  unnest(ip_date)%>%
  unique()

Antibiotics_dates <- split(Antibiotics_dates, Antibiotics_dates$group)

# summary(Antibiotics_dates)
# subset(Antibiotics_dates, group == 'Vanc')%>%
#   left_join(Admission_Scaffold, .)%>%
#   mutate(Vanc = !is.na(group))%>%
#   select(-group)%>%
#   View()


Antibiotics_dates <- sapply(names(Antibiotics_dates), function(xx){names(Antibiotics_dates[[xx]])[3] <- xx
  #browser()
  Antibiotics_dates[[xx]]
  },simplify = FALSE)%>%
  Reduce(left_join, ., Admission_Scaffold)

#mutate(Antibiotics_dates,treatment = if_else(is.na(Other),'',Other))%>%
#  View()

mutate(Antibiotics_dates, across(all_of(c('Other', 'Vanc', 'Zosyn')),~coalesce(.x,'')),
       Exposure = paste(Vanc, Zosyn, Other))%>%
  select(hadm_id, Exposure)%>%
  unique()%>%
  pull(Exposure)%>%
  table()

#class October 5th 2022

Cr_labevents2 <- Antibiotics_dates %>%
  group_by(hadm_id)%>%
  #mutate(Vanc_Zosyn_Date = min(ip_date[!is.na(Vanc) & !is.na(Zosyn)]))%>%
  summarise(Vanc_Zosyn_Date = min(ip_date[!is.na(Vanc) & !is.na(Zosyn)]))%>%
  subset(!is.infinite(Vanc_Zosyn_Date)) %>%
  left_join(Cr_labevents,.)%>%
  group_by(hadm_id)%>%
  subset(!is.na(hadm_id)) %>%
  arrange(hadm_id,charttime)%>%
  mutate(Vanc_Zosyn = !all(is.na(Vanc_Zosyn_Date)))


#combining demographics and creatinine tables
Analysis_Data <- left_join(Cr_labevents2,Demographics1)

ggplot(Analysis_Data, aes(x = Vanc_Zosyn, y = valuenum)) +
  geom_violin()

paireed_analysis <- c('valuenum', 'admits', 'flag', 'Vanc_Zosyn')

Analysis_Data[,paireed_analysis]%>%
  ggpairs(aes(col = Vanc_Zosyn))
#table(Antibiotics_dates$group)%>%
#  View()

table1(~valuenum+admits+flag+anchor_age+gender|Vanc_Zosyn, data = Analysis_Data)


my.render.cont <- function(x) {
  with(stats.default(x),
       sprintf("%0.2f (%0.1f)", MEAN, SD))
}

# subset(Antibiotics, is.na(starttime) & is.na(endtime))
# | is or




group_by(Antibiotic_Groupings, Vanc, Zosyn, Other) %>%
  summarise(N =n())

#debug = {browser();TRUE})



grepl("Zosyn", Antibiotics$label)

export

File: dataprep.py

#start off with reticulate::repl_python() to load interactive session of python in R
import pandas as pd
import numpy as np

import os
import pickle

import re

if not os.path.exists('data.pickle'):
  import runpy
  runpy.run_path('data.py')
  
dd = pickle.load(open('data.pickle', 'rb'))

# dd.keys()
# dd['admissions']

demographics = dd['admissions'].copy()

patients = dd['patients'].copy().drop('dod', axis = 1)

demographics['LOS'] = (pd.to_datetime(demographics['dischtime']) - pd.to_datetime(demographics['admittime']))/np.timedelta64(1, 'D')

demographics1 = demographics.groupby('subject_id').agg(admits = ('subject_id', 'count'),
  eth = ('ethnicity', 'nunique'),
  ethnicity_combo = ('ethnicity', lambda xx: ':'.join(sorted(list(set(xx))))),
  language = ('language', 'last'),
  dod = ('deathtime', lambda xx: max(pd.to_datetime(xx))),
  los = ('LOS', np.median),
  numED = ('edregtime', lambda xx: xx.notnull().sum())).reset_index(drop = False).merge(patients, on = 'subject_id')
  

kw_abx = ["vanco", "zosyn", "piperacillin", "tazobactam", "cefepime", "meropenam", "ertapenem", "carbapenem", "levofloxacin"]
kw_lab = "creatinine"
kw_aki = ["acute renal failure", "acute kidney injury", "acute kidney failure", "acute kidney", "acute renal insufficiency"]
kw_aki_pp = ["postpartum", "labor and delivery"]

label_abx = "|".join(kw_abx)
label_aki = "|".join(kw_aki)
label_aki_pp = "|".join(kw_aki_pp)

items_abx = dd['d_items'][dd['d_items'].label.str.contains(label_abx, case = False)].copy()
items_aki = dd['d_icd_diagnoses'][dd['d_icd_diagnoses'].long_title.str.contains(label_aki, case = False)]
items_aki = items_aki[~items_aki.long_title.str.contains(label_aki_pp, case = False)]
items_lab = dd['d_labitems'][dd['d_labitems'].label.str.contains(kw_lab, case = False, na = False)][items_lab.fluid =="Blood"]
given_abx = items_abx.merge(dd['inputevents'], on = 'itemid')

given_abx['group'] = np.where(given_abx.label=='Vancomycin','Vanc', 
  np.where(given_abx.label.str.contains('Piperacillin'), 'Zosyn', 'Other'))
  
given_abx['starttime'] = pd.to_datetime(given_abx['starttime']).dt.date
given_abx['endtime'] = pd.to_datetime((given_abx['endtime'])).dt.date

given_abx['ip_dates'] = given_abx.apply(lambda row:
  pd.date_range(row['starttime'], row['endtime'], freq = 'D'), axis = 1 )
abx_dates = given_abx.explode('ip_dates')

abx_dates['Vanc'] = 1
abx_dates['Zosyn'] = 1
abx_dates['Other'] = 1



aki_diagnosis = dd['diagnoses_icd'][dd['diagnoses_icd'].icd_code.str.contains("^584|^N17", case = False)]

cr_labevents = items_lab.merge(dd['labevents'], on = 'itemid')
cr_labevents['ip_dates'] = pd.to_datetime(cr_labevents['charttime']).dt.date
cr_labevents = cr_labevents[(cr_labevents.category == 'Chemistry') & (cr_labevents.fluid == 'Blood')][['hadm_id', 'ip_dates','valuenum','flag']].groupby(['hadm_id', 'ip_dates']).agg(
  Creatinine = ('valuenum', 'max'), cr_flag = ('flag', lambda xx: max(np.where(xx == 'abnormal', '1', '0')))).reset_index(drop = False).drop_duplicates()

emar_abx = dd['emar'][dd['emar'].medication.str.contains(label_abx, case = False, na = False)]

#dd['admissions'][['hadm_id', 'admittime', 'dischtime']]

admissions_scaffold = dd['admissions'][['hadm_id', 'admittime', 'dischtime']].copy()
admissions_scaffold['admittime'] = pd.to_datetime(admissions_scaffold['admittime']).dt.round('D')
admissions_scaffold['dischtime'] = pd.to_datetime(admissions_scaffold['dischtime']).dt.round('D')

admissions_scaffold['ip_dates'] = admissions_scaffold.apply(lambda row:
  pd.date_range(row['admittime'], row['dischtime'], freq = 'D'), axis = 1 )
admissions_scaffold = admissions_scaffold.explode('ip_dates')[['hadm_id','ip_dates']]

abx_dates = admissions_scaffold.merge(abx_dates[abx_dates['group'] == 'Zosyn'][['hadm_id','ip_dates', 'Zosyn']], 
  how = 'left').merge(abx_dates[abx_dates['group'] == 'Vanc'][['hadm_id','ip_dates', 'Vanc']], 
  how = 'left').merge(abx_dates[abx_dates['group'] == 'Other'][['hadm_id','ip_dates', 'Other']], 
  how = 'left').fillna(0).drop_duplicates()

abx_dates['ip_dates'] = pd.to_datetime(abx_dates['ip_dates']).dt.date
analysis_data = abx_dates.merge(cr_labevents, on = ['hadm_id', 'ip_dates'], how = 'left').fillna(method = 'ffill', axis = 1)

#start of code for lecture on 11/30/22
#check the previous code and add what was missing from last time




